master/internal/db/postgres_experiments.go

```diff
@@ -1117,65 +1117,48 @@ WITH const AS (
-                                ) v
-                               ) AS validation
-                        FROM steps s
-                        WHERE s.total_batches = c.steps_completed AND s.trial_id = c.trial_id
-                    ) s
-                   ) AS step,
-                   -- We later filter out any checkpoints with any corresponding warm start
-                   -- trials, so we can just put an empty list here. (TODO(dzhu): This is
-                   -- here for backwards compatibility with Python, but could maybe be
-                   -- removed.)
-                   '[]'::jsonb AS warm_start_trials
-            FROM checkpoints_view c, trials t, const
-            WHERE c.state = 'COMPLETED' AND c.trial_id = t.id AND t.experiment_id = $1
-        ) _, const
-    ) c, const
-    WHERE (SELECT COUNT(*) FROM trials t WHERE t.warm_start_checkpoint_id = c.id) = 0
-          AND c.trial_order_rank > $4
-          AND ((c.experiment_rank > $2
-                AND c.trial_rank > $3)
-               OR (c.step->'validation'->'metrics'->'validation_metrics'->>const.metric_name
-                   IS NULL))
-)
-SELECT selected_checkpoints.uuid AS ID from selected_checkpoints;`
+	SELECT *,
+		-- The order includes the id to prevent different rows from having the same
+		-- rank, which could cause more than the desired number of checkpoints to be
+		-- left out of the result set. Also, any rows with null validation values
+		-- will sort to the end, thereby not affecting the ranks of rows with
+		-- non-null validations, and will be filtered out later.
+		rank() OVER (
+			ORDER BY
+				const.sign * (metrics->'validation_metrics'
+							->>const.metric_name)::float8 ASC NULLS LAST, val_id ASC
+		) AS experiment_rank,
+		rank() OVER (
+			PARTITION BY trial_id
+			ORDER BY
+				const.sign * (metrics->'validation_metrics'
+							->>const.metric_name)::float8 ASC NULLS LAST, val_id ASC
+		) AS trial_rank,
+		rank() OVER (
+			PARTITION BY trial_id
+			ORDER BY total_batches DESC
+		) AS trial_order_rank,
+		metrics->'validation_metrics'->>const.metric_name AS val_metric
+	FROM (
+		SELECT ct.*, v.id as val_id, v.metrics,
+			-- We later filter out any checkpoints with any corresponding warm start
+			-- trials, so we can just put an empty list here. (TODO(dzhu): This is
+			-- here for backwards compatibility with Python, but could maybe be
+			-- removed.)
+			'[]'::jsonb AS warm_start_trials
+		FROM (
+			SELECT c.id as c_id, c.trial_id, c.state,
+				c.steps_completed as total_batches, c.report_time as end_time, c.uuid, 
+				c.resources, c.metadata
+			FROM checkpoints_view c, trials t
+			WHERE c.state = 'COMPLETED' AND c.trial_id = t.id AND t.experiment_id = $1 
+		) as ct
+		LEFT JOIN validations v ON v.trial_id = ct.trial_id AND v.total_batches = ct.total_batches
+	) AS sub, const
+	WHERE (SELECT COUNT(*) FROM trials t WHERE t.warm_start_checkpoint_id = sub.c_id) = 0)
```

Comment:
it doesn't make sense to me that making a checkpoint a warm start checkpoint would affect how all the others are GC'd due to `rank()` changes. can we move this to the `WHERE` attached to `FROM selected_checkpoints`.